### PR TITLE
context.xml configuration

### PR DESCRIPTION
**Pull request by bryanjbelanger**
This contains new manifests for context.xml configuration as well as a resource link.

### CODE REVIEW

File: manifests/config/context.pp [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+# - $class_name_ensure specifies whether you are trying to set or remove the
+#   className attribute. Valid values are 'true', 'false', 'present', or
+#   'absent'. Defaults to 'present'.
+# - $address is the optional address attribute.
+# - $address_ensure specifies whether you are trying to set of remove the
+#   address attribute. Valid values are 'true', 'false', 'present', or
+#   'absent'. Defaults to 'present'.
+# - The $port to wait for shutdown commands on.
+# - The $shutdown command that must be sent to $port.
+define tomcat::config::context (
```
Comment (mhaskel) [inline manifests/config/context.pp]: See #66. I think this should probably have the option of either being in server.xml or in a separate file. The implementation in that PR has more flexibility also.

File: manifests/config/context.pp [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+# - $class_name is the optional className attribute.
+# - $class_name_ensure specifies whether you are trying to set or remove the
+#   className attribute. Valid values are 'true', 'false', 'present', or
+#   'absent'. Defaults to 'present'.
+# - $address is the optional address attribute.
+# - $address_ensure specifies whether you are trying to set of remove the
+#   address attribute. Valid values are 'true', 'false', 'present', or
+#   'absent'. Defaults to 'present'.
+# - The $port to wait for shutdown commands on.
+# - The $shutdown command that must be sent to $port.
```
Comment (mhaskel) [inline manifests/config/context.pp]: Comments are copied from a different file

File: manifests/config/context.pp [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+define tomcat::config::context (
+  $catalina_base           = $::tomcat::catalina_home,
+  $class_name              = undef,
+  $class_name_ensure       = 'present',
+  $address                 = undef,
+  $address_ensure          = 'present',
+  $port                    = undef,
+  $shutdown                = undef,
+) {
```
Comment (mhaskel) [inline manifests/config/context.pp]: These parameters aren't actually used.

File: manifests/config/context/resource.pp [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+#   Connector element. Valid values are 'true', 'false', 'present', and
+#   'absent'. Defaults to 'present'.
+# - The $port attribute. This attribute is required unless $connector_ensure
+#   is set to false.
+# - The $protocol attribute. Defaults to $name when not specified.
+# - $parent_service is the Service element this Connector should be nested
+#   beneath. Defaults to 'Catalina'.
+# - An optional hash of $additional_attributes to add to the Connector. Should
+#   be of the format 'attribute' => 'value'.
+# - An optional array of $attributes_to_remove from the Connector.
```
Comment (mhaskel) [inline manifests/config/context/resource.pp]: These comments are all copied from a different file and not accurate

File: manifests/config/context/resource.pp [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+  $auth,
+  $type,
+  $driverClassName,
+  $username,
+  $password,
+  $maxActive,
+  $maxIdle,
+  $maxWait,
+  $url,
+  $factory,
```
Comment (mhaskel) [inline manifests/config/context/resource.pp]: It doesn't look like all these parameters are required. Anything that isn't actually required for the configuration of the Resource elements should be optional.

File: manifests/config/context/resourcelink.pp [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+#   Connector element. Valid values are 'true', 'false', 'present', and
+#   'absent'. Defaults to 'present'.
+# - The $port attribute. This attribute is required unless $connector_ensure
+#   is set to false.
+# - The $protocol attribute. Defaults to $name when not specified.
+# - $parent_service is the Service element this Connector should be nested
+#   beneath. Defaults to 'Catalina'.
+# - An optional hash of $additional_attributes to add to the Connector. Should
+#   be of the format 'attribute' => 'value'.
+# - An optional array of $attributes_to_remove from the Connector.
```
Comment (mhaskel) [inline manifests/config/context/resourcelink.pp]: These comments are not accurate.

File: manifests/config/context/resourcelink.pp [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+define tomcat::config::context::resourcelink (
+  $global,
+  $resource_link_name,
+  $type,
+  $catalina_base   = $::tomcat::catalina_home,
+  $resource_ensure = 'present',
```
Comment (mhaskel) [inline manifests/config/context/resourcelink.pp]: This does not cover all the optional parameters for resource link elements

File: manifests/config/context/resourcelink.pp [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+#   'absent'. Defaults to 'present'.
+# - The $port attribute. This attribute is required unless $connector_ensure
+#   is set to false.
+# - The $protocol attribute. Defaults to $name when not specified.
+# - $parent_service is the Service element this Connector should be nested
+#   beneath. Defaults to 'Catalina'.
+# - An optional hash of $additional_attributes to add to the Connector. Should
+#   be of the format 'attribute' => 'value'.
+# - An optional array of $attributes_to_remove from the Connector.
+define tomcat::config::context::resourcelink (
```
Comment (mhaskel) [inline manifests/config/context/resourcelink.pp]: preference for this being named `tomcat::config::context::resource_link`

File: manifests/config/server/globalnamingresources.pp [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+#   Connector element. Valid values are 'true', 'false', 'present', and
+#   'absent'. Defaults to 'present'.
+# - The $port attribute. This attribute is required unless $connector_ensure
+#   is set to false.
+# - The $protocol attribute. Defaults to $name when not specified.
+# - $parent_service is the Service element this Connector should be nested
+#   beneath. Defaults to 'Catalina'.
+# - An optional hash of $additional_attributes to add to the Connector. Should
+#   be of the format 'attribute' => 'value'.
+# - An optional array of $attributes_to_remove from the Connector.
```
Comment (mhaskel) [inline manifests/config/server/globalnamingresources.pp]: These comments are inaccurate

File: manifests/config/server/globalnamingresources.pp [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+define tomcat::config::server::globalnamingresources (
+  $resource_name,
+  $auth,
+  $type,
+  $factory,
```
Comment (mhaskel) [inline manifests/config/server/globalnamingresources.pp]: It doesn't look like all of these parameters are required for GlobalNamingResource elements

File: manifests/config/server/globalnamingresources.pp [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+  $description     = undef,
+  $driverClassName = undef,
+  $maxActive       = undef,
+  $maxIdle         = undef,
+  $maxWait         = undef,
```
Comment (mhaskel) [inline manifests/config/server/globalnamingresources.pp]: These parameters should have underscores in the name not be camel cased for consistency

**Comment by mhaskel** [comment]
@bryanjbelanger I've made some more specific comments inline, but for more general comments:

* I think ideally this PR would be broken up into several PRs as it's adding a bunch of new functionality that may be easier to merge in smaller pieces
* Each new class/define you add requires documentation updates (top of file and README)
* Each new class/define you add requires at least a new unit test
* Only parameters that are actually required per the tomcat documentation should be required in the manifests
* If implementing elements with several optional parameters considering doing something like https://github.com/puppetlabs/puppetlabs-tomcat/blob/master/manifests/config/server/connector.pp#L24-L25
* If any of these elements are supposed to be nested under parent elements make sure the parents are configurable
* PRs should be rebased/squashed down to at most one commit per feature (including tests/docs for that feature).

Let us know if you have any questions or need help addressing the comments.

**Comment by bryanjbelanger** [comment]
Once I make changes how do I request a new pull?



On Jan 2, 2015, at 6:47 PM, Morgan Haskel <notifications@github.com<mailto:notifications@github.com>> wrote:


@bryanjbelanger<https://github.com/bryanjbelanger> I've made some more specific comments inline, but for more general comments:

  *   I think ideally this PR would be broken up into several PRs as it's adding a bunch of new functionality that may be easier to merge in smaller pieces
  *   Each new class/define you add requires documentation updates (top of file and README)
  *   Each new class/define you add requires at least a new unit test
  *   Only parameters that are actually required per the tomcat documentation should be required in the manifests
  *   If implementing elements with several optional parameters considering doing something like https://github.com/puppetlabs/puppetlabs-tomcat/blob/master/manifests/config/server/connector.pp#L24-L25
  *   If any of these elements are supposed to be nested under parent elements make sure the parents are configurable
  *   PRs should be rebased/squashed down to at most one commit per feature (including tests/docs for that feature).

Let us know if you have any questions or need help addressing the comments.

-

**Comment by mhaskel** [comment]
@bryanjbelanger you can either close this PR and just submit a new PR, or you can use rebasing with squashing and forced updates to your branch.

In general we recommend using the [Github Flow workflow](https://guides.github.com/introduction/flow/index.html)

**Comment by bryanjbelanger** [comment]
Need to pull and resubmit.

**Pull request closed without merging**